### PR TITLE
Generic File Parser: Locations & Accessions

### PR DESCRIPTION
**Pull request by dwaring87**
Description 
-----------------------------------------------------

This adds the `CXGN::File::Parse` class that can be used to parse any supported data file format into a uniform parsed data format (see #4975 for details).  It also updated the Locations and Accessions upload to use the new class.

Checklist 
---------------------------------------------------------------------------------------
- [ ] Refactoring only
- [ ] Documentation only
- [ ] Fixture update only
- [ ] Bug fix
  - [ ] The relevant issue has been closed.
  - [ ] Further work is required.
- [ ] New feature
  - [ ] Relevant tests have been created and run.
  - [ ] Data was added to the fixture
    - [ ] Data was added via a patch in `/t/data/fixture/patches/`.
  - [ ] User-Facing Change
    - [ ] The user manual in `/docs` has been updated.
  - [ ] Any new Perl has been documented using **perldoc**.
  - [ ] Any new JavaScript has been documented using **JSDoc**.
  - [ ] Any new _legacy_ JavaScript has been moved from `/js` to `/js/source/legacy`.

### CODE REVIEW

File: lib/CXGN/File/Parse/Plugin/Excel.pm [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      $v = $super->clean_value($v, $hv);
+      $row_info{$hv} = $v;
+
+      if ( $v && $v ne '' ) {
+        if ( ref($v) eq 'ARRAY' ) {
+          if ( scalar(@$v) > 0 ) {
+            foreach (@$v) {
+              $values_map{$hv}->{$_} = 1;
+            }
+            $skip_row = 0;
```
Comment (titima15) [inline lib/CXGN/File/Parse/Plugin/Excel.pm:78]: Users have had issues when they manipulated excel files. Sometimes the files ended up having no data at the end of the files, but they are not actually empty. As a result, the loop is not ended. It helps after I changed from skip to last.

File: lib/CXGN/File/Parse.pm [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+  my $data = $parsed->{data};
+  my $values = $parsed->{values};
+  my $required_columns = $parsed->{required_columns};
+  my $opional_columns = $parsed->{optional_columns};
+  my $additional_columns = $parsed->{additional_columns};
+
+COLUMN ALIASES
+  - Alternate column headers can be specified as column_aliases
+  - This will replace the alternate name with the preferred name in the final parsed data
+
```
Comment (titima15) [inline lib/CXGN/File/Parse.pm:88]: Do we really need column aliases?

**Comment by titima15** [comment]
I think it's very nice to have a generic file parser. The only draw back is that it aggregates data based on each row. Then we need to go over each row again to re-structure the data for validation step, depending on the type of data. It should be fine, just a bit slower, I guess.

**Review by titima15 — APPROVED** [review]
👍

**Comment by dwaring87** [comment]
Changes from recent commits:
- stop parsing a file (Excel or Plain) if there are 5 blank rows in a row
- remove line on spreadsheet info about columns can be in any order (could be confusing to users)

I also tested a large accessions template (with 5000 entries).  The original parsing & verification step took 5.66 minutes, the generic parsing & verification took 5.69 minutes (so equally slow in both versions).

**Review by nmenda — APPROVED** [review]

**Pull request merged**